### ISSUE TITLE
Python version and other environment questions

### ISSUE THREAD

**Issue by hdoupe** [OpenSourcePolicyCenter/PolicyBrain#773]
While resetting my local environment, I got the following error:
```
Install webapp-public requirements.txt
Collecting Django==1.8.2 (from -r ../requirements.txt (line 1))
/anaconda/envs/aei_dropq/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:318: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/anaconda/envs/aei_dropq/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Could not fetch URL https://pypi.python.org/simple/django/: There was a problem confirming the ssl certificate: [Errno 1] _ssl.c:504: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed - skipping
  Could not find a version that satisfies the requirement Django==1.8.2 (from -r ../requirements.txt (line 1)) (from versions: )
No matching distribution found for Django==1.8.2 (from -r ../requirements.txt (line 1))
FAILED
~/Documents/PolicyBrain
```

This occurred when setting requirements in the [`requirements.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/7da856c1740023ab245bf46e5a1863d60748da10/requirements.txt) file from this function in [`install_taxbrain_server.sh`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/deploy/install_taxbrain_server.sh#L19-L20).

Eventually, this led to the realization that the `aei_dropq` environment python version was downgraded from python 2.7.12 (set in [`dropq_environment.yml`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/deploy/fab/dropq_environment.yml)) to python 2.7.3 when a package named [`pillow`](https://python-pillow.org/) was installed. 

```
install -c ospc/label/dev -c ospc pillow
Fetching package metadata ...............
Solving package specifications: .

Package plan for installation in environment /anaconda/envs/aei_dropq:

The following NEW packages will be INSTALLED:

    freetype: 2.8-h12048fb_1      
    jpeg:     9b-he5867d9_2       
    libpng:   1.6.32-hd1e8b91_4   
    libtiff:  4.0.9-h0dac147_0    
    olefile:  0.44-py27h73ba740_0 
    pillow:   4.3.0-py27hb485eaf_1
    xz:       5.2.3-h0278029_2    

The following packages will be UPDATED:

    tk:       8.5.18-0             --> 8.6.7-h35a86e2_3

The following packages will be DOWNGRADED:

    python:   2.7.12-1             --> 2.7.3-5         
```

I removed the `pillow` package from the [`requirements.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/7da856c1740023ab245bf46e5a1863d60748da10/requirements.txt) file and the local set up worked just fine.  As expected, the python version at the end was 2.7.12.  Pillow is an image processing library and I cannot find where it is used in the app or why it would downgrade the python version.  Also, `pillow` was added in the [initial commit](https://github.com/OpenSourcePolicyCenter/PolicyBrain/commit/7a9c6fc493a0f7c7d387cb536885c768a6d8f8a2).

Going through this prompts several questions:
1. Most immediately, do we use pillow and can it be removed?
2. How many other packages are there that we do not need anymore?
3. Why do we have 6 files for setting the environment?  How can we simplify this?
    - [`deploy/fab/dropq_environment.yml`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/deploy/fab/dropq_environment.yml)
    - [`deploy/requirements.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/deploy/requirements.txt)
    - [`conda-requirements.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/conda-requirements.txt)
    - [`requirements.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/requirements.txt)
    - [`requirements_dev.txt`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/requirements_dev.txt)
    - Other EC2 environment setup occurs in [`reset_server.sh`](https://github.com/OpenSourcePolicyCenter/PolicyBrain/blob/master/deploy/fab/reset_server.sh). 
4. The AWS EC2 nodes use python 2.7.13, and the heroku app appears to be on 2.7.14.  From a previous environment export, I know that the local development environment is on 2.7.12, but today, it was bumped back to 2.7.3.  How important is it that all three use the same version of python?

@talumbau if you have any insight that you can add, then that would be very helpful.  Even if you only have time to hit one or two of these questions, that would be greatly appreciated.

**Comment by hdoupe** [comment]
Answer to 1. Pillow is required and has something to do with imagefields.  See:
```
(aei_dropq) HDoupe-MacBook-Pro:PolicyBrain henrydoupe$ python manage.py makemigrations
SystemCheckError: System check identified some issues:

ERRORS:
hermes.Post.hero: (fields.E210) Cannot use ImageField because Pillow is not installed.
	HINT: Get Pillow at https://pypi.python.org/pypi/Pillow or run command "pip install Pillow".
```

**Comment by talumbau** [comment]
> 
> Most immediately, do we use pillow and can it be removed?
> How many other packages are there that we do not need anymore?

This seems like a hard one to answer and, as long as there aren't issues with slug sizes on Heroku, it is probably ok to allow the installation of packages that are not strictly needed to run TaxBrain.

> Why do we have 6 files for setting the environment? How can we simplify this?
> deploy/fab/dropq_environment.yml
> deploy/requirements.txt
> conda-requirements.txt
> requirements.txt
> requirements_dev.txt
> Other EC2 environment setup occurs in reset_server.sh.

The history is as follows: 
`requirements.txt` - Heroku environments are typically specified as pip-installable packages listed in a requirements.txt. 

`conda-requirements.txt` - This started out as a set of packages that would be installed in to the Heroku environment via conda instead of pip because if we specified things like `numpy` in the `requirements.txt` it would trigger a compile step on the numpy source code, which was obviously bad.

`requirements_dev.txt` - This was a nice place to specify things like `pytest`, which aren't needed to run the Heroku environment, but are needed to do dev work on it.

`deploy/fab/dropq_environment.txt` - This was originally the conda environment for the worker nodes.

`deploy/requirements.txt` - I don't know about this one.

`reset_server.sh` - This script was originally intended to solve the following problem: a worker node is up and ready to receive workloads for TaxBrain, CCC, etc. A new version of `taxcalc`, `btax`, etc. is now available we want to 1. stop the web server momentarily 2. upgrade to the latest version of `taxcalc`, `btax`, etc., 3. start serving again. Maybe the right answer here is just to do proper salt states instead of this (now quite complex!) script?

 
> The AWS EC2 nodes use python 2.7.13, and the heroku app appears to be on 2.7.14. From a previous environment export, I know that the local development environment is on 2.7.12, but today, it was bumped back to 2.7.3. How important is it that all three use the same version of python?

I think you could probably get away with something like `python >= 2.7.12` in specifying the Python version for various environments. Python 2.7.3 is probably too old though. I would suggest making updates to the environment specification files to prevent a regression of the fix you mention in this issue.

**Comment by hdoupe** [comment]
Thanks @talumbau.  Your insight is helpful.  I'll be working through these issues today.

Ok, now I see.  I'd always that the heroku environment, the ec2 node environment, and the local environment were the same.  Through @talumbau's explanation and working through this problem I now understand that they are different and are setup with different processes.

The worker node environment and the heroku app environment being different makes sense since they have different objectives and constraints.  However, the local environment is created by merging the two environments.  This is done by running the 5 or 6 scripts above sequentially.  Since they are run as separate processes, there is no guarantee that requirements set at step 1 are still adhered to in later steps.  

I think a more robust solution would be to have two (maybe 3) explicit environments, created via environment.yml files and requirements.txt files as needed.  Environment 1 would be for the front-end app which is analogous to our heroku environment now.  

Environment 2 would be for the worker nodes.  As @talumbau suggested, this should probably be set in the salt files.  This would also guarantee that non-OSPC packages are updated during each deployment.  For example, pandas is on version 0.20.1 which was released in May, and numpy is on version 0.12.1 which was released in March.  Note that as more models are added to PolicyBrain it could make sense to have worker nodes operating in different environments for different models as long as some basic requirements are met.

Environment 3 would be the local environment.  The current local development setup is a merge of environments 1 and 2.  The local environment setup routine seems to require this.  In an ideal world, this discrepancy between the local and production environments would not be needed.

For now, we can solve the version problem, but in the spring, we should do some work on cleaning up the environment setup processes.

**Comment by talumbau** [comment]
Happy to help. One item of note is that we never had a concept of
"environment 3". I would always just maintain two separate console sessions
with environment 1 for running Django and environment 2 for running the
worker node process (and a console session for running reddis).

**Comment by hdoupe** [comment]
@talumbau Ok, I see. That makes sense.  I think the project has morphed into having an environment 3.  Currently, everything is run in the `aei_dropq` environment for the local setup.  One console runs the webapp.  The other runs supervisord which manages redis, celery, and flask.

**Comment by hdoupe** [comment]
Having a very different environment 3 is definitely suboptimal.  In the spring, we have several large projects planned for PolicyBrain.  Fixing this will be one of them.

**Comment by hdoupe** [comment]
I'm closing this in favor of a comprehensive issue on PolicyBrain deployment and environment setup.

**Comment by hdoupe** [comment]
Closed #773
